### PR TITLE
feat: Add programs for creating savedata & bootstrap keys

### PR DESCRIPTION
**Pull request by nurupo**
A few programs to help with savedata creation. They can create savedata using a provided secret key, e.g. from `cracker.c`, as well as random savedata (not cracking, just random).

`create_minimal_savedata.c` replaces `minimal-save-generator.c` because it's the same `minimal-save-generator.c` but with the ability to use a provided key.

`create_savedata.c` is the same as `create_minimal_savedata.c` but using Tox API instead of hacking savedata together. It requires `-ltoxcore` and generates bigger savedata.

`create_bootstrap_keys.c ` is the same but for bootstrap node keys.

---
This change is [Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/1818)

### CODE REVIEW

File: other/fun/create_savedata.c [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+static bool print_savedata(Tox *const tox)
+{
+    const size_t savedata_size = tox_get_savedata_size(tox);
+    uint8_t *const savedata = (uint8_t *const)malloc(savedata_size);
+
+    if (savedata == nullptr) {
+        tox_kill(tox);
+        return false;
```
Comment (iphydf) [inline other/fun/create_savedata.c:50]: `Memory leak: savedata` -- cppcheck

File: other/fun/create_savedata.c [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    }
+
+    bin2hex_toupper(tox_id_str, tox_id_str_size, tox_id, tox_address_size());
+    fprintf(stdout, "Tox Id: %s\n", tox_id_str);
+    free(tox_id_str);
+    free(tox_id);
+    return true;
+}
+
+int main(const int argc, const char *const argv[])
```
Comment (iphydf) [inline other/fun/create_savedata.c:83]: Add this to other/analysis/gen-file.sh:

```sh
echo "static void call(int m(int, const char *const *), int argc, char **argv) { m(argc, argv); }" >>amalgamation.cc
```

**Review by iphydf — CHANGES_REQUESTED** [review]

**Review by JFreegman — COMMENTED** [review]
> Reviewed 6 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/1818)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @nurupo)

___
*[other/fun/create_common.h, line 20 at r1](https://reviewable.io/reviews/toktok/c-toxcore/1818#-MsAxpm94eebAuPqsLPy:-MsAxpm94eebAuPqsLPz:b-g6lt3f) ([raw file](https://github.com/toktok/c-toxcore/blob/679d2871028cdec572e573f66c7e9eb4363c644b/other/fun/create_common.h#L20)):*
> ```objc
> }
> 
> static int tox_strcasecmp(const char *s1, const char *s2)
> ```

Why not use the strcasecmp in strings.h?
___
*[other/fun/create_minimal_savedata.c, line 37 at r1](https://reviewable.io/reviews/toktok/c-toxcore/1818#-MsAxGgE8gCUUUuqdH-A:-MsAxGgE8gCUUUuqdH-B:b-ms2em2) ([raw file](https://github.com/toktok/c-toxcore/blob/679d2871028cdec572e573f66c7e9eb4363c644b/other/fun/create_minimal_savedata.c#L37)):*
> ```C
> 
>     for (size_t i = 0; i < crypto_box_PUBLICKEYBYTES; i ++) {
>         checksum[i % 2] ^= public_key[i];
> ```

Should probably use a defined value/constant here instead of `2` since it should always be tied to the size of checksum

File: other/fun/create_savedata.c (hunk 1 above)
Comment (nurupo) [reply other/fun/create_savedata.c:50]: Are you sure? `savedata` doesn't leak, it's `NULL` here.

The issue here is the `tox_kill(tox);` call I forgot to remove after all the refactoring I did, resulting in a double-kill. Fixing that now...

**Review by nurupo — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/1818)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @iphydf and @JFreegman)

___
*[other/fun/create_common.h, line 20 at r1](https://reviewable.io/reviews/toktok/c-toxcore/1818#-MsAxpm94eebAuPqsLPy:-MsB2q2c0bvlHcDUOSCA:bo54g8r) ([raw file](https://github.com/toktok/c-toxcore/blob/679d2871028cdec572e573f66c7e9eb4363c644b/other/fun/create_common.h#L20)):*
<details><summary><i>Previously, JFreegman wrote…</i></summary><blockquote>

Why not use the strcasecmp in strings.h?
</blockquote></details>

Those are non-standard C functions. They are only in the POSIX standard, so not available on Windows.
___
*[other/fun/create_minimal_savedata.c, line 37 at r1](https://reviewable.io/reviews/toktok/c-toxcore/1818#-MsAxGgE8gCUUUuqdH-A:-MsB67azAGCd-U9IsDlP:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/679d2871028cdec572e573f66c7e9eb4363c644b/other/fun/create_minimal_savedata.c#L37)):*
<details><summary><i>Previously, JFreegman wrote…</i></summary><blockquote>

Should probably use a defined value/constant here instead of `2` since it should always be tied to the size of checksum
</blockquote></details>

Done.
___
*[other/fun/create_savedata.c, line 83 at r1](https://reviewable.io/reviews/toktok/c-toxcore/1818#-Ms1dpnmWr0HddjgPjTJ-r1-83:-MsB68xx8eYARqcMUc-_:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/679d2871028cdec572e573f66c7e9eb4363c644b/other/fun/create_savedata.c#L83)):*
<details><summary><i>Previously, iphydf wrote…</i></summary><blockquote>

Add this to other/analysis/gen-file.sh:

```sh
echo "static void call(int m(int, const char *const *), int argc, char **argv) { m(argc, argv); }" >>amalgamation.cc
```
</blockquote></details>

Done.

**Comment by nurupo** [comment]
___
*[other/fun/create_savedata.c, line 50 at r1](https://reviewable.io/reviews/toktok/c-toxcore/1818#-Ms1dpnmWr0HddjgPjTJ-r1-50:-MsB6eeW--jsLWwDwMK1:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/679d2871028cdec572e573f66c7e9eb4363c644b/other/fun/create_savedata.c#L50)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

Are you sure? `savedata` doesn't leak, it's `NULL` here.

The issue here is the `tox_kill(tox);` call I forgot to remove after all the refactoring I did, resulting in a double-kill. Fixing that now...
</blockquote></details>

Done.

**Review by JFreegman — APPROVED** [review]
> Reviewed 1 of 2 files at r3, 1 of 3 files at r4, 4 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/1818)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @iphydf)

**Comment by nurupo** [comment]
```
$ git push origin master 
warning: not sending a push certificate since the receiving end does not support --signed push
Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
remote: error: GH006: Protected branch update failed for refs/heads/master.
remote: error: Required status check "license/cla" is pending.
To github.com:TokTok/c-toxcore.git
 ! [remote rejected]   master -> master (protected branch hook declined)
error: failed to push some refs to 'github.com:TokTok/c-toxcore.git'
```

@iphydf could your merge this PR?

**Pull request merged**